### PR TITLE
[URP] Automatic Alpha-To-Coverage Support

### PR DESCRIPTION
**Pull request by gmitrano-unity**
### Purpose of this PR
This change attempts to automatically enable alpha-to-coverage
functionality whenever alpha clipping is in use on opaque materials
in URP. Since this functionality relies on MSAA, it is only active for
forward passes. With this change, alpha clipped geometry such as
foliage should see improved visual quality when MSAA is enabled.

No MSAA:
image

4x MSAA:
image

No MSAA (5x Zoom):
image

4x MSAA (5x Zoom):
image

---
### Testing status
Manual testing on PC on Viking Village with MSAA on and off with both forward and deferred rendering modes.
Yamato with new AlphaToCoverage Graphics Tests
TODO: Mobile testing

---
### Notes for Reviewers
I believe I've handled all of the edge cases at this point, but I'd appreciate extra review scrutiny on the material override handling and possible consequences of rendering out non 1.0 alpha values in cases where we weren't previously. This should only ever happen during opaque rendering though.

### Notes for QA
This change is intended to improve the edge quality of alpha-clipped (ex: foliage) geometry **only** when MSAA is enabled.
It **should** have no visual effect when MSAA is off. The visual quality is also expected to change/improve with the number of MSAA samples (ex: 2x, 4x, 8x)

### CODE REVIEW

**Comment by LaserYGD** [comment]
Does this already exist for HDRP's MSAA?

**Comment by sebastienlagarde** [comment]
FYI, following the proposal in this PR to have alpha to coverage always enabled when  MSAA is on, we have done the change in HDRP: https://github.com/Unity-Technologies/Graphics/pull/6599

**Comment by gmitrano-unity** [comment]
> FYI, following the proposal in this PR to have alpha to coverage always enabled when MSAA is on, we have done the change in HDRP: #6599 (before this changed , alpha to coverage was manually enabled in HDRP)

Took me a while to understand why the change is so much simpler in HDRP, but it's because of the special alpha testing logic in the depth prepass.  Due to that, HDRP only needs to mess with a2c state for the prepass which means the normal lit shader pass doesn't have to deal with any edge cases of having a2c on or off. Nice!

Unfortunately, I don't think URP can do the same because it doesn't guarantee a depth prepass. That leaves us with something a bit more complex, but at least it'll be automated for both pipelines soon! 🙂

File: com.unity.render-pipelines.universal/ShaderLibrary/UnityInput.hlsl [hunk 1]
```diff
@@ -63,6 +63,9 @@ float4 _ScreenParams;
 // y = 2.0 ^ [Mip Bias]
 float2 _GlobalMipBias;
 
+// 1.0 if AlphaToMask is enabled and 0.0 otherwise
+float _AlphaToMaskEnabled;
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/ShaderLibrary/UnityInput.hlsl:67]: Is this matching with built-in? I think the UnityInput was for values from the Engine, or values that we setup but match built-in/engine.

File: com.unity.render-pipelines.universal/Shaders/BakedLit.shader [hunk 2]
```diff
@@ -41,6 +42,8 @@ Shader "Universal Render Pipeline/Baked Lit"
             Name "BakedLit"
             Tags{ "LightMode" = "UniversalForwardOnly" }
 
+            AlphaToMask[_AlphaToMask]
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Shaders/BakedLit.shader:45]: I wonder if this is actually useful for BakedLit?

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl [hunk 3]
```diff
@@ -378,7 +378,10 @@ half4 SpeedTree8Frag(SpeedTreeFragmentInput input) : SV_Target
     half4 diffuse = SampleAlbedoAlpha(uv, TEXTURE2D_ARGS(_MainTex, sampler_MainTex)) * _Color;
 
     half alpha = diffuse.a * input.interpolated.color.a;
-    AlphaDiscard(alpha, 0.3333);
+
+#if defined(_ALPHATEST_ON)
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl:382]: Previously the `#if defined(_ALPHATEST_ON)` was in the inside of the function to not require explicit `#ifdef` soup outside.
I personally like it explicit, so you instantly know if it's done or not, but it's inconsistent how many similar functions work in URP shaders.
Also potentially breaking user code (bu that's already broken by the API change.)

File: com.unity.render-pipelines.universal/Shaders/Terrain/WavingGrass.shader [hunk 4]
```diff
@@ -16,6 +16,8 @@ Shader "Hidden/TerrainEngine/Details/UniversalPipeline/WavingDoublePass"
 
         Pass
         {
+            AlphaToMask On
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Shaders/Terrain/WavingGrass.shader:19]: Should this be `AlphaToMask[_AlphaToMask]`?

**Review by eh-unity — COMMENTED** [review]
Some review comments about API breakage and URP shader style inconstencies.
Missing changelog.

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/Includes/PBRForwardPass.hlsl [hunk 5]
```diff
@@ -70,14 +70,19 @@ half4 frag(PackedVaryings packedInput) : SV_TARGET
...
-    #if _ALPHATEST_ON
-        half alpha = surfaceDescription.Alpha;
-        clip(alpha - surfaceDescription.AlphaClipThreshold);
-    #elif _SURFACE_TYPE_TRANSPARENT
-        half alpha = surfaceDescription.Alpha;
-    #else
-        half alpha = 1;
-    #endif
+#if defined(_SURFACE_TYPE_TRANSPARENT)
+    half surfaceType = 1.0;
```
[the remark this answers is not in the document]
Comment (gmitrano-unity) [reply com.unity.render-pipelines.universal/Editor/ShaderGraph/Includes/PBRForwardPass.hlsl]: I replaced this type of logic with a simpler isTransparent bool, but I also added constants for surface types as well.

File: com.unity.render-pipelines.universal/ShaderLibrary/ShaderVariablesFunctions.hlsl [hunk 6]
```diff
@@ -149,17 +149,51 @@ void GetLeftHandedViewSpaceMatrices(out float4x4 viewMatrix, out float4x4 projMa
...
+bool IsAlphaToMaskEnabled()
 {
-    #ifdef _ALPHATEST_ON
-    if (IsAlphaDiscardEnabled())
-        clip(alpha - cutoff + offset);
-    #endif
+    return (_AlphaToMaskEnabled != 0.0);
+}
+
+half AlphaClip(half alpha, half cutoff)
```
[the remark this answers is not in the document]
Comment (gmitrano-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/ShaderVariablesFunctions.hlsl:185]: 1. I added the old function back and piped it through into this new logic and it cuts down the diff a bit too which is a nice side effect.
2. The offset parameter wasn't used anywhere and all we did with it before was add it into the cutoff parameter so it didn't seem very useful to me. It's still available in the AlphaDiscard() function in my latest change.

File: com.unity.render-pipelines.universal/ShaderLibrary/ShaderVariablesFunctions.hlsl [hunk 7]
```diff
@@ -149,17 +149,51 @@ void GetLeftHandedViewSpaceMatrices(out float4x4 viewMatrix, out float4x4 projMa
     projMatrix._13_23_33_43 = -projMatrix._13_23_33_43;
 }
 
-void AlphaDiscard(real alpha, real cutoff, real offset = real(0.0))
+// Only define the alpha clipping helpers when the alpha test define is present.
+// This should help identify usage errors early.
+#if defined(_ALPHATEST_ON)
```
[the remark this answers is not in the document]
Comment (gmitrano-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/ShaderVariablesFunctions.hlsl:172]: Restoring the AlphaDiscard() reduces the impact of this issue a bit, but overall I'm not sure why we would want to expose these functions to users in cases where they cannot function as intended? The previous implementation just turns into a no-op when the alpha test macro isn't defined. This seems like a case of silent failure versus loud error and I usually prefer loud errors myself. :)

If we were to expose these functions, I think their implementation would just always return false for IsAlphaToMaskAvailable(), and always return input alpha for AlphaClip(). Not sure if that would be useful?

File: com.unity.render-pipelines.universal/ShaderLibrary/UnityGBuffer.hlsl [hunk 8]
```diff
@@ -134,7 +134,7 @@ FragmentOutput SurfaceDataToGbuffer(SurfaceData surfaceData, InputData inputData
     output.GBuffer0 = half4(surfaceData.albedo.rgb, PackMaterialFlags(materialFlags));   // albedo          albedo          albedo          materialFlags   (sRGB rendertarget)
     output.GBuffer1 = half4(surfaceData.specular.rgb, surfaceData.occlusion);            // specular        specular        specular        occlusion
     output.GBuffer2 = half4(packedNormalWS, surfaceData.smoothness);                     // encoded-normal  encoded-normal  encoded-normal  smoothness
-    output.GBuffer3 = half4(globalIllumination, 1);                                      // GI              GI              GI              [optional: see OutputAlpha()] (lighting buffer)
+    output.GBuffer3 = half4(globalIllumination, 1);                                      // GI              GI              GI              unused          (lighting buffer)
```
[the remark this answers is not in the document]
Comment (gmitrano-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/UnityGBuffer.hlsl:137]: I actually removed this because I did look at OutputAlpha() based on the comment and I didn't find anything about how it interacted with the GBuffer. Not sure exactly what the intention was there.

My changes do actually affect the output alpha in some cases, but none of those cases should involve GBuffers. AlphaToMask requires MSAA so my changes are supposed to be limited to forward rendering only.

File: com.unity.render-pipelines.universal/ShaderLibrary/UnityInput.hlsl (hunk 1 above)
Comment (gmitrano-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/UnityInput.hlsl]: Hmmm, no it's not matching built-in. I was just looking for a place I could put a shader parameter where it'd be globally accessible. Is there a better place for this type of thing in URP?

File: com.unity.render-pipelines.universal/ShaderLibrary/ShaderVariablesFunctions.hlsl (hunk 7 above)
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/ShaderVariablesFunctions.hlsl:172]: I agree with you as I mentioned. I like to fail fast too. 

But in the past I've been told to not do this, so that we don't couple our users to our keywords (perhaps due to variant explosion). 
Users can set the constants from code, so it is functional code, but of course they can set keywords too. Or if they fork the shader or SRP they can just add a define.
Let me get another opinion on it and maybe we can agree to change it.

We could also put the Alpha stuff  in it's own file and you just include that if you need Alpha ops.

File: com.unity.render-pipelines.universal/ShaderLibrary/UnityInput.hlsl (hunk 1 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/UnityInput.hlsl]: Looking at the UnityInput.hlsl, it looks like the "rule" has been broken already. 
UnityInput.hlsl was supposed to match 
https://docs.unity3d.com/2022.1/Documentation/Manual/SL-UnityShaderVariables.html
 as far as I know, but there's other stuff too like _GlobalMipBias.
 Perhaps, it's not an issue then.

I think the correct place would be Input.hlsl where all the URP constants are.

File: com.unity.render-pipelines.universal/ShaderLibrary/ShaderVariablesFunctions.hlsl (hunk 6 above)
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/ShaderVariablesFunctions.hlsl:185]: 1. That should be good enough.
2. I agree, it doesn't seem very useful, but I would've thought something used it. Anyway it's in the public API.

File: com.unity.render-pipelines.universal/ShaderLibrary/ShaderVariablesFunctions.hlsl (hunk 7 above)
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/ShaderVariablesFunctions.hlsl:172]: We discussed this internally. It should be ok to land like the way you have it now.

**Review by aleksandrasdzikia — APPROVED** [review]
LGTM. Found issue but not related to this PR. Other than that results are as expected.

**Comment by sebastienlagarde** [comment]
@gmitrano-unity guess this PR have landed in private repo? can you close it then? thnaks
